### DISCUSSION TITLE
How to add lock file in deno.json?

### DISCUSSION THREAD

**Discussion by DrSensor** [denoland/deno#15915] [Q&A]
Hi, is there a way to specify the lock file so that all tasks (`deno task`) in _deno.json_ will use that lock file automatically?

Or at least I can specify it in the command line explicitly for specific task.
For example: `deno --lock=lock.json task build`

<details><summary>my config</summary>

---
deno.json
```json
{
  "importMap": "import_map.json",
  "tasks": {
    "lume": "deno eval \"import 'lume/task.ts'\" --",
    "build": "deno task lume",
    "serve": "deno task lume -s"
  },
  "compilerOptions": {
    "jsx": "react-jsx",
    "jsxImportSource": "preact",
    "types": [
      "./types.d.ts"
    ],
    "lib": [
      "dom",
      "dom.iterable",
      "dom.asynciterable",
      "deno.ns",
      "deno.unstable"
    ]
  }
}
```
---
import_map.json
```json
{
  "imports": {
    "preact/jsx-runtime": "https://esm.sh/preact/jsx-runtime",
    "preact/jsx-dev-runtime": "https://esm.sh/preact/jsx-dev-runtime",
    "lume/": "https://deno.land/x/lume/"
  }
}
```
</details>

**Comment by zachauten** [comment]
[The `--lock` flag is only available on commands that populate the cache](https://deno.land/manual@v1.25.3/getting_started/command_line_interface#integrity-flags-lock-files). `task` isn't one of these. But if you want a task that uses one of the deno commands that _does_ use the `--lock` flag, you should be able to:
```
"tasks": {
  "cache-with-lock": "deno cache --lock=lock.json deps.ts"
}
```

**Reply by DrSensor** [reply]
Can I use it in `deno eval`?
For example (taken from [Lume](https://lume.land/docs/overview/installation/#don't-install-lume) on how to use the CLI)
```
deno eval "import 'lume/task.ts'"
```
My setup is kinda problematic since I can't specify the dependencies in deps.ts plus I want dependencies used by Lume subcommands (including Lume itself) respect the lockfile.

**Reply by zachauten** [reply]
No, deno eval doesn’t write to the cache either. It sounds like maybe you are just trying to vendor lume? Does [installing](https://lume.land/docs/overview/installation/#install-lume-globally-on-your-computer) it work for your use case? Or maybe [deno vendor](https://deno.land/manual@v1.25.4/tools/vendor)?

**Reply by DrSensor** [reply]
Installing it works but the update method is based on pinned version in the import_map.json. I think `deno vendor` works for me since it support vendoring from import-map. Can it vendor the whole path? I mean is it going to traverse all files and dirs in the prefix url? For example,
```json
{
  "imports": {
    "plugins/": "https://esm.sh/lume/plugins/"
  }
}
```
will that download all `https://esm.sh/lume/plugins/*.ts` and `https://esm.sh/lume/plugins/**/*.ts`?

**Reply by DrSensor** [reply]
I think I will go with 1 file that only consist of this single line
```ts
/// lume.js
import "lume/task.ts"
```
then I can do `deno run --lock=lock.json lume.js`. Kinda ugly but at least it works on my favor.

**Reply by DrSensor** [reply]
Does `deno cache` and `--lock-write` also detect dynamic import?

**Reply by zachauten** [reply]
> Does deno cache and --lock-write also detect dynamic import?

I'm not sure. Try running deno cache on a file with a dynamic import and check if it writes anything to the lockfile?
